// get-window-offset/build/cjs/src/nearWhere.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getNearLocation = exports["default"] = void 0;

var _getDomCenter = _interopRequireDefault(require("./getDomCenter"));

var _getDomPositionInfo2 = _interopRequireDefault(require("./getDomPositionInfo"));

var getNearLocation = function getNearLocation(center, floatInfo) {
  var loc = {
    center: false,
    centerCenter: false,
    top: false,
    bottom: false,
    left: false,
    right: false
  };

  if (floatInfo.x > center.x) {
    loc.right = true;
  } else if (floatInfo.x < center.x) {
    loc.left = true;
  } else {
    loc.center = true;
  }

  if (floatInfo.y > center.y) {
    loc.bottom = true;
  } else if (floatInfo.y < center.y) {
    loc.top = true;
  } else {
    loc.center = true;
  }

  if (loc.center) {
    if (!loc.top && !loc.bottom && !loc.left && !loc.right) {
      loc.centerCenter = true;
    }
  }

  return loc;
};

exports.getNearLocation = getNearLocation;

var nearWhere = function nearWhere(targetEl, floatEl) {
  var tarCenter = (0, _getDomCenter["default"])(targetEl);
  var floatXY;

  if (floatEl.nodeName) {
    var _getDomPositionInfo;

    var floatElInfo = ((_getDomPositionInfo = (0, _getDomPositionInfo2["default"])(floatEl)) === null || _getDomPositionInfo === void 0 ? void 0 : _getDomPositionInfo.domInfo) || {
      top: 0,
      left: 0
    };
    floatXY = {
      x: floatElInfo.left,
      y: floatElInfo.top
    };
  }

  if (null == floatXY) {
    floatXY = floatEl;
  }

  return getNearLocation({
    x: tarCenter[0],
    y: tarCenter[1]
  }, floatXY);
};

var _default = nearWhere;
exports["default"] = _default;
