// reshow-flux/build/cjs/useConnect.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("reshow-runtime/helpers/slicedToArray"));

var _react = require("react");

var _reshowHooks = require("reshow-hooks");

var _reshowConstant = require("reshow-constant");

var _getStore = _interopRequireDefault(require("./getStore"));

var handleShouldComponentUpdate = function handleShouldComponentUpdate(_ref) {
  var options = _ref.options,
      shouldComponentUpdate = _ref.shouldComponentUpdate,
      calculateState = _ref.calculateState,
      prev = _ref.prev,
      props = _ref.props;
  var nextState = calculateState(prev.state, options);
  var bUpdate = !shouldComponentUpdate || shouldComponentUpdate({
    prev: prev,
    nextProps: props,
    nextState: nextState
  });

  if (!bUpdate || props === prev.props && nextState === prev.state) {
    prev.__init__ = _reshowConstant.T_TRUE;
    return prev;
  } else {
    return {
      props: props,
      __init__: _reshowConstant.T_TRUE,
      state: nextState
    };
  }
};

var useConnect = function useConnect() {
  var inputOptions = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return function (props) {
    var options = (0, _getStore["default"])({
      options: inputOptions,
      props: props
    });
    var calculateState = options.calculateState,
        shouldComponentUpdate = options.shouldComponentUpdate,
        _options$displayName = options.displayName,
        displayName = _options$displayName === void 0 ? "useConnect" : _options$displayName;
    (0, _react.useDebugValue)(displayName);

    var _useState = (0, _react.useState)(function () {
      return {
        props: props,
        state: calculateState({}, options)
      };
    }),
        _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
        data = _useState2[0],
        setData = _useState2[1];

    var isMount = (0, _reshowHooks.useMounted)();
    (0, _react.useEffect)(function () {
      var handleChange = function handleChange() {
        var storeState = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : options.storeState;

        if (_reshowConstant.T_TRUE === isMount()) {
          options.storeState = storeState;
          setData(function (prev) {
            return handleShouldComponentUpdate({
              options: options,
              shouldComponentUpdate: shouldComponentUpdate,
              calculateState: calculateState,
              prev: prev,
              props: props
            });
          });
        }
      };

      if (!data.__init__ || data.props !== props) {
        handleChange();
      }

      options.store.addListener(handleChange);
      return function () {
        options.store.removeListener(handleChange);
      };
    }, props.changeable ? [props] : []);
    return data.state || {};
  };
};

var _default = useConnect;
exports["default"] = _default;
module.exports = exports.default;
