// react-atomic-molecule/build/cjs/ui/molecules/Form.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("reshow-runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("reshow-runtime/helpers/objectSpread2"));

var _objectWithoutProperties2 = _interopRequireDefault(require("reshow-runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _classLib = require("class-lib");

var _SemanticUI = _interopRequireDefault(require("../molecules/SemanticUI"));

var _useCSS = _interopRequireDefault(require("../../src/useCSS"));

var _excluded = ["messageType", "style", "className"];

var Form = function Form(props) {
  (0, _useCSS["default"])(["input", "search", "form"], "semantic");
  var messageType = props.messageType,
      style = props.style,
      className = props.className,
      others = (0, _objectWithoutProperties2["default"])(props, _excluded);
  var classes = (0, _classLib.mixClass)(className, messageType, "form");
  return /*#__PURE__*/_react["default"].createElement(_SemanticUI["default"], (0, _extends2["default"])({
    atom: "form"
  }, others, {
    className: classes,
    style: (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, Styles.form), style)
  }));
};

var _default = Form;
exports["default"] = _default;
var Styles = {
  form: {
    boxSizing: "border-box"
  }
};
module.exports = exports.default;
