// need-css/build/es/src/index.js
import { css } from "create-el";
import { win } from "win-doc";
import get from "get-object-value"; // https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/components/

var SEMANTIC_VERSION = "/npm/semantic-ui@2.4.2/dist/components";
var semanticMap = {
  button: "".concat(SEMANTIC_VERSION, "/button.min.css"),
  card: "".concat(SEMANTIC_VERSION, "/card.min.css"),
  checkbox: "".concat(SEMANTIC_VERSION, "/checkbox.min.css"),
  dimmer: "".concat(SEMANTIC_VERSION, "/dimmer.min.css"),
  divider: "".concat(SEMANTIC_VERSION, "/divider.min.css"),
  dropdown: "".concat(SEMANTIC_VERSION, "/dropdown.min.css"),
  form: "".concat(SEMANTIC_VERSION, "/form.min.css"),
  header: "".concat(SEMANTIC_VERSION, "/header.min.css"),
  image: "".concat(SEMANTIC_VERSION, "/image.min.css"),
  input: "".concat(SEMANTIC_VERSION, "/input.min.css"),
  item: "".concat(SEMANTIC_VERSION, "/item.min.css"),
  label: "".concat(SEMANTIC_VERSION, "/label.min.css"),
  list: "".concat(SEMANTIC_VERSION, "/list.min.css"),
  menu: "".concat(SEMANTIC_VERSION, "/menu.min.css"),
  message: "".concat(SEMANTIC_VERSION, "/message.min.css"),
  modal: "".concat(SEMANTIC_VERSION, "/modal.min.css"),
  popup: "".concat(SEMANTIC_VERSION, "/popup.min.css"),
  progress: "".concat(SEMANTIC_VERSION, "/progress.min.css"),
  rail: "".concat(SEMANTIC_VERSION, "/rail.min.css"),
  search: "".concat(SEMANTIC_VERSION, "/search.min.css"),
  segment: "".concat(SEMANTIC_VERSION, "/segment.min.css"),
  tab: "".concat(SEMANTIC_VERSION, "/tab.min.css"),
  table: "".concat(SEMANTIC_VERSION, "/table.min.css")
};
var cssMap = {
  "fixed-data-table": "/npm/fixed-data-table-2@0.8.13/dist/fixed-data-table.min.css",
  prismjs: "/npm/prismjs/themes/prism.min.css"
};
var group = {
  semantic: semanticMap,
  "default": cssMap
};
var cdn = "https://cdn.jsdelivr.net";
var load = [];
/**
 * Clean bfcache for offline
 */

win().addEventListener && win().addEventListener("load", function () {
  return load.splice(0, load.length);
});

var needCSS = function needCSS(mods) {
  var groupKey = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "default";

  if (win().__null) {
    return;
  }

  var map = get(group, [groupKey], {});
  mods.forEach(function (mod) {
    var url = get(map, [mod]);
    url = url ? cdn + url : mod;

    if (url) {
      if (load[url]) {
        return;
      } else {
        load[url] = true;
        css()()(url);
      }
    }
  });
};

export default needCSS;
