// call-func/build/es/src/index.js
import callfunc from "./callfunc";

var defaultCall = function defaultCall(defaultFunc, func, scope) {
  return function () {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return callfunc(func || defaultFunc, args, scope);
  };
};

var isRequired = function isRequired() {
  var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "param";
  throw new Error("".concat(name, " is required"));
};
/**
 * only run lasttime.
 */


var debounce = function debounce(func, defaultDelay) {
  var timer;
  return function (option) {
    var _ref = option || {},
        _ref$delay = _ref.delay,
        delay = _ref$delay === void 0 ? 250 : _ref$delay,
        args = _ref.args,
        scope = _ref.scope;

    clearTimeout(timer);
    timer = setTimeout(function () {
      return callfunc(func, args, scope);
    }, defaultDelay || delay);
  };
};
/**
 * reduce run times.
 */


var throttle = function throttle(func) {
  var threshhold = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 250;
  var needRunLast = arguments.length > 2 ? arguments[2] : undefined;
  var waiting = false;
  var lastCall = false;

  var run = function run(option) {
    var _ref2 = option || {},
        args = _ref2.args,
        scope = _ref2.scope;

    lastCall = false;
    callfunc(func, args, scope);
  };

  return function (option) {
    lastCall = true;

    if (!waiting) {
      waiting = true;
      run(option);
      setTimeout(function () {
        waiting = false;

        if (needRunLast && lastCall) {
          run(option);
        }
      }, threshhold);
    }
  };
};

export default callfunc;
export { defaultCall, isRequired, debounce, throttle };
export { default as register, cleanAllRegister } from "./register";
