// react-atomic-molecule/build/es/src/lib/styles/index.js
import _typeof from "reshow-runtime/es/helpers/typeof";
import ucfirst from "ucfirst-js";
import StyleObject from "./StyleObject";
import store from "./store";
import nToU from "./cssNumberToUnit";
import { IS_ARRAY, KEYS, UNDEFINED } from "reshow-constant";
var Browser = {
  webkit: "Webkit",
  ms: "ms",
  moz: "Moz"
};

var genStyleId = function genStyleId() {
  store.counter += 1;
  return "c" + store.counter + "_";
};

var createStyle = function createStyle(css, selector, styleId) {
  if (!css) {
    return;
  }

  if (UNDEFINED === _typeof(styleId)) {
    styleId = genStyleId();
  } else if (store.registry[styleId]) {
    return store.registry[styleId];
  }

  if (!IS_ARRAY(css)) {
    css = [css];
  }

  var styles = [];
  css.forEach(function (one, i) {
    styles[i] = {};
    KEYS(one).forEach(function (key) {
      if (IS_ARRAY(one[key]) && 1 === one[key].length) {
        var ucFirstKey = ucfirst(key);
        styles[i][Browser.webkit + ucFirstKey] = styles[i][Browser.ms + ucFirstKey] = styles[i][Browser.moz + ucFirstKey] = styles[i][key] = nToU(key, one[key][0]);
      } else {
        styles[i][key] = nToU(key, one[key]);
      }
    });
  });
  var styleDecl = new StyleObject(styles, selector, styleId);
  store.newStyles.push(styleDecl);
  return styleDecl;
};

export default createStyle; // reactStyle(css, selector, styleId)
