// react-atomic-molecule/build/es/src/lib/styles/injectStyle.js
import { doc } from "win-doc";
import { inject, create } from "create-el";
import query from "css-query-selector";
import { KEYS } from "reshow-constant";
import store from "./store.js";
import stylesToCSS from "./stylesToCSS";

var appendCss = function appendCss(cssArr) {
  KEYS(cssArr).forEach(function (key) {
    var id = "react-style-" + key;
    var styleDom = query.one("#" + id);
    var css = cssArr[key];

    if (styleDom) {
      styleDom.innerHTML = css;
    } else {
      styleDom = create("style")()({
        id: id,
        innerHTML: css
      });
      inject(function () {
        return doc().getElementsByTagName("head")[0];
      })(styleDom);
    }
  });
};

var injectStyle = function injectStyle() {
  if (!store.newStyles.length) {
    // We are in Node or Styles are already injected
    return null;
  }

  var compiled = stylesToCSS(store.newStyles);
  store.lastCompiled = compiled;
  store.newStyles = [];
  KEYS(compiled.styleIds).forEach(function (key) {
    return store.registry[key] = compiled.styleIds[key];
  });

  if (compiled.css) {
    if (doc().URL) {
      appendCss(compiled.cssArr);
    } else {
      console.log(compiled.css);
    }
  }
};

export default injectStyle;
