// get-object-value/build/cjs/src/index.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toMap = exports.toJS = exports.initMap = exports.getDefault = exports["default"] = void 0;

var _typeof2 = _interopRequireDefault(require("reshow-runtime/helpers/typeof"));

var _reshowConstant = require("reshow-constant");

var getWebpack4Default = function getWebpack4Default(o) {
  return get(o, [_reshowConstant.DEFAULT, _reshowConstant.DEFAULT], function () {
    return get(o, [_reshowConstant.DEFAULT], function () {
      return o;
    });
  });
};

exports.getDefault = getWebpack4Default;

var getDefaultValue = function getDefaultValue(v, cur) {
  return _reshowConstant.FUNCTION === (0, _typeof2["default"])(v) ? v(cur) : v !== null && v !== void 0 ? v : cur;
};

var toJS = function toJS(v) {
  return v && v.toJS ? v.toJS() : v;
};

exports.toJS = toJS;

var toMap = function toMap(a, path) {
  var next = get(a, path, {});
  var nextMap = {};
  (0, _reshowConstant.KEYS)(next).forEach(function (key) {
    return nextMap[key] = toJS(next[key]);
  });
  return nextMap;
};

exports.toMap = toMap;

var initMap = function initMap(o) {
  return function (k, defaultValue) {
    return o[k] || (o[k] = getDefaultValue(defaultValue));
  };
};

exports.initMap = initMap;

var get = function get(o, path, defaultValue) {
  if (null == o) {
    return getDefaultValue(defaultValue, o);
  }

  var current = toJS(o);

  if (!path || !(0, _reshowConstant.IS_ARRAY)(path)) {
    return current;
  }

  try {
    var i = path.length;
    var j = path.length - 1;

    while (i--) {
      var index = path[j - i];

      if (current && _reshowConstant.UNDEFINED !== (0, _typeof2["default"])(current[index])) {
        current = current[index];

        if (null == current) {
          current = getDefaultValue(defaultValue, current);
          break;
        }
      } else {
        current = getDefaultValue(defaultValue);
        break;
      }
    }
  } catch (e) {
    console.warn({
      e: e
    });
    current = getDefaultValue(defaultValue);
  }

  return current;
};

var _default = get;
exports["default"] = _default;
