// react-atomic-molecule/build/es/ui/molecules/Form.js
import _extends from "reshow-runtime/es/helpers/extends";
import _objectSpread from "reshow-runtime/es/helpers/objectSpread2";
import _objectWithoutProperties from "reshow-runtime/es/helpers/objectWithoutProperties";
var _excluded = ["messageType", "style", "className"];
import React from "react";
import { mixClass } from "class-lib";
import SemanticUI from "../molecules/SemanticUI";
import useCSS from "../../src/useCSS";

var Form = function Form(props) {
  useCSS(["input", "search", "form"], "semantic");

  var messageType = props.messageType,
      style = props.style,
      className = props.className,
      others = _objectWithoutProperties(props, _excluded);

  var classes = mixClass(className, messageType, "form");
  return /*#__PURE__*/React.createElement(SemanticUI, _extends({
    atom: "form"
  }, others, {
    className: classes,
    style: _objectSpread(_objectSpread({}, Styles.form), style)
  }));
};

export default Form;
var Styles = {
  form: {
    boxSizing: "border-box"
  }
};
