// smooth-scroll-to/build/cjs/src/index.js
"use strict";

var _interopRequireDefault = require("reshow-runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _getScrollInfo = require("get-scroll-info");

var _callFunc = _interopRequireDefault(require("call-func"));

var _easingLib = require("easing-lib");

var isRunning = false;
/**
 *  !!Important!! any logic change need take care isRunning
 */

var smoothScrollTo = function smoothScrollTo(to, duration, el, callback) {
  var scrollNode = (0, _getScrollInfo.getScrollNode)(el);

  var cb = function cb() {
    isRunning = false;
    (0, _callFunc["default"])(callback, [scrollNode]);
  };

  if (isRunning) {
    isRunning = false;
    setTimeout(function () {
      scrollNode.scrollTop = to;
      cb();
    });
    return false;
  }

  if (!duration) {
    duration = 900;
  }

  var from = scrollNode.scrollTop;
  var go = to - from;

  if (!go) {
    return cb();
  }

  (0, _easingLib.aniTimer)({
    isContinue: function isContinue(elapsedTime) {
      var progress = (0, _easingLib.easeInOutCubic)(elapsedTime, from, go, duration);
      scrollNode.scrollTop = progress;

      if (isRunning) {
        return true;
      }
    },
    cancel: function cancel() {
      return cb();
    },
    done: function done() {
      return cb();
    }
  }, duration);
  isRunning = true;
};

var _default = smoothScrollTo;
exports["default"] = _default;
module.exports = exports.default;
